src/frequenz/dispatch/_bg_service.py

```diff
@@ -23,6 +27,77 @@
 """The logger for this module."""
 
 
+class _MergeStrategy(ABC):
+    """Base class for strategies to merge running intervals."""
+
+    @abstractmethod
+    def _get_filter_function(
+        self,
+        scheduler: DispatchScheduler,
+    ) -> Callable[[Dispatch], bool]:
+        """Get a filter function for dispatches.
+
+        Args:
+            scheduler: The dispatch scheduler.
+
+        Returns:
+            A filter function.
+        """
```

Comment:
This should all be public, an important reason to uses classes is so that it is easy to extend from outside the library itself!

Reply:
Also, why `get_filter_function()` and not just `filter()`?

```suggestion
class MergeStrategy(ABC):
    """Base class for strategies to merge running intervals."""

    @abstractmethod
    def filter(self, dispatch: Dispatch) -> bool:
        """Get a filter function for dispatches.

        Args:
            scheduler: The dispatch scheduler.

        Returns:
            A filter function.
        """
```

And if any strategy needs the scheduler[^1], it should take it in the constructor.

```py
class MergeByType(_MergeStrategy):
    """Merge running intervals based on the dispatch type."""

    def __init__(self, scheduler: DispatchScheduler) -> None:
        """Initialize the strategy."""
        self._scheduler: DispatchScheduler = scheduler
```

It also feels to me that a merge strategy should be able to provide a `map()` function too (for example you might want to strip or add stuff from the `Dispatch` and return a different type), but maybe this is not really necessary :thinking: 

[^1]: Maybe we should expose the list of dispatches directly in the dispatcher, or expose the scheduler, to make creating a merge strategy more high-level.

Reply:
I felt, requiring the user to explicitly pass the scheduler was unnecessary work demand on them. I guess we could still do it but provide shortcuts through `Dispatcher.mergeByType()` and similar